core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -301,6 +306,7 @@ public synchronized void schedule()
                                     });
                                 }
                                 if (taskSource.isFinished()) {
+                                    dynamicFilterService.stageCannotScheduleMoreTasks(stage.getStageId(), 0, allPartitions.size());
```

Comment:
My intention is to collect dynamic filters from the worker only after the task is successful (see usage of fetchFinalDynamicFiltersIfNecessary in HttpRemoteTask). I'm assuming that each logical task will be successful only once and so report DF only once.
The `all` case can indeed be reported before task completion, but that case has special handling in DynamicFilterService anyway to not wait for other tasks if any task reports an `all`. Please let me know if you still see a loophole in that logic.

Reply:
> I'm assuming that each logical task will be successful only once and so report DF only once.

This assumption may be true today. However in the future we may need to schedule multiple instances of the same task speculatively in case of a suspected failure (by default it takes up to 5 minuted to consider a lost task as failed, we may prefer to speculatively schedule an another instance after some timeout).

Reply:
I've updated PR now with changes that don't depend on task FINISH for collecting DF